### DISCUSSION TITLE
spacy v3 pretrain / static vectors ndim mismatch

### DISCUSSION THREAD

**Discussion by adamkgoldfarb** [explosion/spaCy#7474] [Help: Best practices]
Hello! 

I'm doing a bit of pretraining before loading up some glove and fasttext vectors for training on CPU. I'm hitting a mismatch when training:

`Attempt to change dimension 'nM' for model 'static_vectors' from 300 to 0`

I haven't changed the config from pretraining to training, but I _did_ change the attrs from ["ORTH","SHAPE"] to ["NORM","PREFIX","SUFFIX","SHAPE"] to align with the recommendations in the documentation. Not sure if that would have any affect, but since I'm not changing the config I assumed it would be ok from pretraining to training.

Going to move ahead without the pretraining bit but wondering if anything springs to mind for you!

Here are the relevant elements of the config, as far as I can tell (let me know if you need more):

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[pretraining]
max_epochs = 1000
dropout = 0.2
n_save_every = null
component = "tok2vec"
layer = ""
corpus = "corpora.pretrain"

[pretraining.batcher]
@batchers = "spacy.batch_by_words.v1"
size = 3000
discard_oversize = false
tolerance = 0.2
get_length = null

[pretraining.objective]
@architectures = "spacy.PretrainCharacters.v1"
maxout_pieces = 3
hidden_size = 300
n_characters = 4

[pretraining.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = true
eps = 0.00000001
learn_rate = 0.001

**Comment by svlandeg** [comment]
Hi Adam, sorry to hear you're running into trouble!

I could be mistaken - but I don't think this is due to the `attrs` you changed. Instead, this error feels like the vectors aren't being found (setting to dim 0). Are your paths all correct?

Could you share the full config file, and the exact command(s) you ran that led to this error? It might also help to share the full stack trace, as that will have additional clues :-)

**Reply by adamkgoldfarb** [reply]
It would make a ton of sense if the vectors weren't being found, or _were_ being found and are just configured incorrectly: I've init-ed vectors from fasttext and glove, and I've gone through training (without the pretraining step) with seven different versions but have not seen any significant variance in best/last scores. 

However, final models _do_ vary in size, indicating that the vectors might still be in there. I'll walk through the steps I took and then post the stack trace and config. In advance: I really appreciate your eyes on this, and fyi that spacy projects have helped keep track of the necessary steps!

## init vectors

`python -m spacy init vectors en vectors/pre-trained_glove/glove.6B.300d.txt vectors/glove/glove.6B.300d`

## pretrain weights

`python -m spacy pretrain configs/config.cfg training/pretrain_weights --paths.raw_text corpus/free_responses.jsonl`

## train model

`python -m spacy train configs/config.cfg --output models/gl-glove-6B-300d-test --paths.train training/lea_annotations/train.spacy --paths.dev training/lea_annotations/dev.spacy --paths.init_tok2vec training/pretrain_weights/model20.bin --paths.vectors vectors/glove/glove.6B.300d --paths.raw_text corpus/free_responses.jsonl`

## stack trace

✔ Created output directory: models/gl-glove-6B-300d-test
ℹ Using CPU

=========================== Initializing pipeline ===========================
[2021-03-18 07:34:38,775] [INFO] Set up nlp object from config
[2021-03-18 07:34:38,787] [INFO] Pipeline: ['tok2vec', 'ner']
[2021-03-18 07:34:38,792] [INFO] Created vocabulary
[2021-03-18 07:34:40,104] [INFO] Added vectors: vectors/glove/glove.6B.300d
[2021-03-18 07:34:40,104] [INFO] Finished initializing nlp object
Traceback (most recent call last):
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/cli/_util.py", line 69, in setup_cli
    command(prog_name=COMMAND)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/cli/train.py", line 56, in train_cli
    nlp = init_nlp(config, use_gpu=use_gpu)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/training/initialize.py", line 71, in init_nlp
    nlp.initialize(lambda: train_corpus(nlp), sgd=optimizer)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/language.py", line 1250, in initialize
    init_tok2vec(self, P, I)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/training/initialize.py", line 159, in init_tok2vec
    layer.from_bytes(weights_data)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/model.py", line 526, in from_bytes
    return self.from_dict(msg)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/model.py", line 549, in from_dict
    node.set_dim(dim, value)
  File "/Users/adamgoldfarb/anaconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/model.py", line 188, in set_dim
    raise ValueError(err)
ValueError: Attempt to change dimension 'nM' for model 'static_vectors' from 300 to 0

## config

[paths]
train = null
dev = null
vectors = null
init_tok2vec = null
raw_text = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = true
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.pretrain]
@readers = "spacy.JsonlCorpus.v1"
path = ${paths.raw_text}
min_length = 5
max_length = 500
limit = 0

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = true

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_per_type = null
ents_f = 0
ents_p = 0.4
ents_r = 0.6

[pretraining]
max_epochs = 1000
dropout = 0.2
n_save_every = null
component = "tok2vec"
layer = ""
corpus = "corpora.pretrain"

[pretraining.batcher]
@batchers = "spacy.batch_by_words.v1"
size = 3000
discard_oversize = false
tolerance = 0.2
get_length = null

[pretraining.objective]
@architectures = "spacy.PretrainCharacters.v1"
maxout_pieces = 3
hidden_size = 300
n_characters = 4

[pretraining.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = true
eps = 0.00000001
learn_rate = 0.001

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.components.ner]

[initialize.components.ner.labels]
@readers = "spacy.read_labels.v1"
path = "training/lea_annotations/labels/ner.json"

[initialize.tokenizer]

**Reply by svlandeg** [reply]
First quick hunch:

I think the `pretrain` command should also get the path to the vectors?

i.e. instead of 
```
python -m spacy pretrain configs/config.cfg training/pretrain_weights --paths.raw_text corpus/free_responses.jsonl
```
try
```
python -m spacy pretrain configs/config.cfg training/pretrain_weights --paths.raw_text corpus/free_responses.jsonl --paths.vectors vectors/glove/glove.6B.300d
```
?

**Reply by adamkgoldfarb** [reply]
Ah! I definitely didn't get that pretraining required the target vectors-- though it's clear now from "Specifically, we load pretrained vectors, and train a component like a CNN, BiLSTM, etc to predict vectors which match the pretrained ones" in [the docs](https://spacy.io/api/cli#pretrain). 

Pretrain accepted the `--paths.vectors` argument. I'll try to train after one epoch and the `vectors/glove/glove.6B.300d` vectors and will report back. If successful, this will mean pretraining once for each vector set-- good to know!

**Reply by adamkgoldfarb** [reply]
Pretraining passing the --paths.vectors argument works! I was able to successfully initiate training with the pretrained model#.bin. A couple of follow-up questions, as I dove back into the documentation to jog my memory as to why I set up pretraining that way:

1. In the documentation for [PretrainVectors](https://spacy.io/api/architectures#pretrain_vectors), one must "make sure that the initialize.vectors section in the config refers to a model with static vectors." So that makes sense-- the objective is based on the static vectors. 
2. The documentation for [PretrainCharacters](https://spacy.io/api/architectures#pretrain_chars) doesn't mention including a static vectors table, but that's the model I was using in the config, and indeed it seems that that's necessary. 

I can test this, but do you anticipate any benefit to using the character pretraining for e.g. subword-based vectors and vector pretraining otherwise?

**Reply by svlandeg** [reply]
That's really difficult to say, it depends on your specific challenge and dataset. I'd be interested in your results though ;-) 

The reason why you need the static vectors is really because of this bit in your config:
```
[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
include_static_vectors = true
...
[initialize]
vectors = ${paths.vectors}
```
When you pass in the argument `--paths.vectors` you're changing the config file by overwriting `paths.vectors`, and thus you're also changing the internal structure of your neural networks. And you need to make sure that these are exactly the same between pretraining and training, or spaCy will notice they're not, and will throw the `Attempt to change dimension` error you saw initially.

**Reply by adamkgoldfarb** [reply]
Interesting results to share! Hopefully these are helpful reflections for the community.

1. I'm using CPU-- it's an 8-Core Intel Core i9 but it's still CPU. I was following default recommendations to go with CPU-based training because that'd allow for faster iteration using a wide variety of configurations. So it has.
2. I used the PretrainVectors architecture and let pretraining run on my relatively small corpus of 18k (longish) texts for 20 epochs. That took a good eight to ten hours per pretraining, one set of weights for each of [these fasttext vectors](https://fasttext.cc/docs/en/english-vectors.html) and [glove vectors](https://nlp.stanford.edu/projects/glove/). I figured I'd only have pretrain once against each vector set, so it was worth it to let the computer huff and puff.
3. I am optimizing for recall (0.6R/0.4P weighting) because this pipeline is leading to an entity linker and we'd rather capture more entities here while maintaining face validity for demonstration purposes.
4. I saw no significant difference in performance in recall or accuracy when using vector-based pretraining + fasttext's subword-based vectors  VS. vector-based pretraining + vectors trained *without* subword information. Recall and Accuracy are pretty high: 93 R and 88.4 P for the fasttext wiki_news_300d_1M vectors with pretraining predicting my one label, which is a pretty tractable label as it aligns fairly well with LOC/GPE labels in other models.
5. Training with subword vectors and the character pretraining objective got us... less than the PretrainVectors architecture! Though not by much in terms of recall, and precision was actually better. I have a feeling my use case is not the right one for character-based pretraining.
6. Training with roberta-base as drop-in replacement tok2vec got us to 95% recall and 91.5% precision in about 6 hours! That compares favorably to overall accuracy in the static vector-based winner, and in less time overall than pretraining plus training. I suppose that's not super surprising-- roberta has been "pretrained" on a huge data set, just not the one we're using.
7. Training with distilroberta-base even passes the "anxious analyst" test by showing us a progress_bar that moves at a noticeable (albeit still slow) clip, taking about 4.5 hrs total. 
8. The size of the resulting roberta-base and distilroberta-base models are also 2-4x smaller than those that need to pass around static fasttext or glove vectors.
9. As you might imagine, performance at runtime using transformers on CPU suffers. See the table below. The overall tradeoff between performance at training and inference time vs. accuracy holds-- in our case though, we have time to sit around and let the transformers do their thing while we go about our business. This is a batch process we're running rather than a real-time one, and I think trading off speed for accuracy is acceptable for us.

| model                            | recall (%) | precision (%) | speed (w/s?) | size (mb) |
| -------------------------------- | ---------- | ------------- | ------------ | --------- |
| lea_glove_6B_300d                | 91.3       | 86.3          | 5720         | 527       |
| lea_ft_wiki_news_300d_1M_subword | 90.5       | 89.1          | 6030         | 1300      |
| lea_ft_wiki_news_300d_1m         | 92.9       | 88.4          | 6002         | 1300      |
| lea_distilroberta_base_trf       | 92.9       | 92.0          | 1619         | 330       |
| lea_roberta_base_trf             | 95.0       | 91.5          | 1005         | 501       |
| lea_distilbert_conll03_trf       | 93.9       | 92.2          | 2206         | 262  
[Model Perfomance]

Just for posterity, the phrase "This requires a word vectors model to be trained and loaded" in the [Pretraining Objectives](https://spacy.io/usage/embeddings-transformers#pretraining-details) section of the docs implied to me that they are required for PretrainVectors but not for PretrainCharacters, though my understanding was incorrect *if I'm using static vectors later on during training*, which I am here.

So I'm clear, are vectors required for the [characters](https://spacy.io/api/architectures#pretrain_chars) objective if I'm *not* using static vectors? In other words, would it make sense to pretrain on the characters objective and use that to init the tok2vec if you're *not* using static vectors as a feature, as long as that's reflected in your config? Thanks for indulging me on this question!

And more broadly, all of the above would not have been possible without the excellent work your team has done to enable rapid experimentation, so thank you, Thank You, THANK YOU!

Edit: added lea_distilbert_conll03_trf based on [Elastic's distilbert model](https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english) finetuned for NER. I should also note that I'm looking forward to finally trying out the entity linker using the prodigy recipe from your tutorial!

**Reply by adamkgoldfarb** [reply]
@svlandeg Let me know if I need to take this to Stack Overflow, as this isn't really the right thread, but: are there plans in the works to adapt [nel-wikipedia](https://github.com/explosion/projects/tree/master/nel-wikipedia) for v3? I think that might be a good exercise for me to take on as I'm diving into wikidata anyway, but don't want to clash if work on that is well underway.

**Reply by svlandeg** [reply]
Hi, new questions can always go in a new topic on this forum ;-)
But no, we're not planning on porting that Wikidata/Wikipedia parsing code to the v3 core library.

**Reply by adamkgoldfarb** [reply]
Good to know-- thanks again!!
